source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+    "arrayFilters": Optional<Array>,
+    "hint": Optional<String>
+}
+```
+
+#### Delete
+
+```json
+{
+    "delete": <Int32>,
+    "filter": <Document>,
+    "multi": <boolean>,
+    "hint": Optional<String>,
+    "collation": Optional<Document>
+}
+```
+
+### Namespace Information
+
+The `nsInfo` field is an array containing the namespaces on which the write operations should be
+performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the
+`nsInfo` array have the following format:
+
+```json
+{
+    "ns": <String>
+}
+```
+
+### `errorsOnly` and `verboseResults`
+
+The `errorsOnly` field indicates whether the results cursor returned in the `bulkWrite` response
+should contain only errors and omit individual results. If false, both individual results for
+successful operations and errors will be returned.
+
+This field corresponds to the `verboseResults` option defined on `BulkWriteOptions`. Its value
+should be provided as `false` if `verboseResults` was set to `true`; otherwise, it should be
+provided as `true`. This field MUST always be defined in the `bulkWrite` command regardless of
+whether the user specified a value for `verboseResults`.
+
+## Command Batching
+
+Drivers MUST accept an arbitrary number of operations as input to the `MongoClient.bulkWrite` method.
+Because the server imposes restrictions on the size of write operations, this means that a single
+call to `MongoClient.bulkWrite` may require multiple `bulkWrite` commands to be sent to the server.
+Drivers MUST split bulk writes into separate commands when the user's list of operations exceeds
+one or more of these maximums: `maxWriteBatchSize`, `maxBsonObjectSize`, and `maxMessageSizeBytes`.
+Each of these values can be retrieved from the selected server's `hello` command response. Drivers
+MUST merge results from multiple batches into a single `BulkWriteResult` or `BulkWriteException`
+to return from `MongoClient.bulkWrite`.
+
+### Number of Writes
+
+The `ops` array MUST NOT include more than `maxWriteBatchSize` operations.
+
+### Total Message Size
+
+#### Encrypted bulk writes
+
+When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
+sequences and MUST limit the size of the command according to the auto-encryption size limits
+defined in the
+[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).
+
+#### Unencrypted bulk writes with document sequences
+
+When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
+of the `OP_MSG` built for the `bulkWrite` command does not exceed `maxMessageSizeBytes`. Some
+drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the
+server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the
+combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted to accommodate for the bytes in
```

Comment:
I'm not sure if other drivers will encounter a similar issue, but in Rust we build all of the operation-specific portions of a command prior to appending operation-agnostic fields and building an actual `OP_MSG`. This means that the total size of the message isn't known at the time of constructing `ops` and `nsInfo`. I lifted the 16 KiB number from the extra bytes the server allows in documents for overhead and it seems to be fine from my testing.

Reply:
Sounds reasonable. Should you clarify that 16KiB is an approximation?